File: app/src/main/java/com/bitflaker/lucidsourcekit/main/dreamjournal/DreamJournalEditorRatingView.java
package com.bitflaker.lucidsourcekit.main.dreamjournal;

import android.content.pm.ActivityInfo;
import android.content.res.ColorStateList;
import android.os.Bundle;
import android.text.TextUtils;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.ImageView;
import android.widget.LinearLayout;
import android.widget.ToggleButton;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.fragment.app.Fragment;

import com.bitflaker.lucidsourcekit.R;
import com.bitflaker.lucidsourcekit.data.enums.journalratings.DreamClarity;
import com.bitflaker.lucidsourcekit.data.enums.journalratings.DreamMoods;
import com.bitflaker.lucidsourcekit.data.enums.journalratings.DreamTypes;
import com.bitflaker.lucidsourcekit.data.enums.journalratings.SleepQuality;
import com.bitflaker.lucidsourcekit.database.dreamjournal.entities.resulttables.DreamJournalEntry;
import com.bitflaker.lucidsourcekit.databinding.FragmentJournalEditorRatingBinding;
import com.bitflaker.lucidsourcekit.utils.Tools;
import com.google.android.material.dialog.MaterialAlertDialogBuilder;

import io.reactivex.rxjava3.disposables.CompositeDisposable;

public class DreamJournalEditorRatingView extends Fragment {
    private final String[] dreamMoodLabels = new String[] { "Terrible", "Poor", "Okay", "Great", "Outstanding"};
    private final String[] sleepQualityLabels = new String[] { "Terrible", "Poor", "Great", "Outstanding"};
    private final String[] dreamClarityLabels = new String[] { "Very Cloudy", "Cloudy", "Clear", "Crystal Clear"};
    private ImageView[] dreamMoods = new ImageView[5];
    private ImageView[] sleepQualities = new ImageView[4];
    private ImageView[] dreamClarities = new ImageView[4];
    private ColorStateList unselectedIconColor;
    private ColorStateList selectedIconColor;
    private CompositeDisposable compositeDisposable;
    private DreamJournalEntry entry;
    private OnBackButtonClicked mBackButtonListener;
    private OnDoneButtonClicked mDoneButtonListener;
    private FragmentJournalEditorRatingBinding binding;
    private OnCloseButtonClicked mCloseButtonClicked;

    @Override
    public View onCreateView(@NonNull LayoutInflater inflater, ViewGroup container, Bundle savedInstanceState) {
        binding = FragmentJournalEditorRatingBinding.inflate(inflater, container, false);
        return binding.getRoot();
    }

    @Override
    public void onViewCreated(@NonNull View view, @Nullable Bundle savedInstanceState) {
        super.onViewCreated(view, savedInstanceState);

        compositeDisposable = new CompositeDisposable();
        unselectedIconColor = Tools.getAttrColorStateList(R.attr.colorSurfaceContainerHigh, getContext().getTheme());
        selectedIconColor = Tools.getAttrColorStateList(R.attr.colorOnSurface, getContext().getTheme());

        dreamMoods = new ImageView[] {
                binding.imgVeryDissatisfied,
                binding.imgDissatisfied,
                binding.imgNeutralSatisfied,
                binding.imgSatisfied,
                binding.imgVerySatisfied
        };
        sleepQualities = new ImageView[] {
                binding.imgVeryBadQuality,
                binding.imgBadQuality,
                binding.imgGoodQuality,
                binding.imgVeryGoodQuality
        };
        dreamClarities = new ImageView[] {
                binding.imgVeryUnclear,
                binding.imgUnclear,
                binding.imgClear,
                binding.imgVeryClear
        };
        restoreStoredEntry();

        binding.tglDjNightmare.setOnClickListener(e -> toggleSpecialDreamType((ToggleButton) e, DreamTypes.Nightmare.getId()));
        binding.tglDjParalysis.setOnClickListener(e -> toggleSpecialDreamType((ToggleButton) e, DreamTypes.SleepParalysis.getId()));
        binding.tglDjLucid.setOnClickListener(e -> toggleSpecialDreamType((ToggleButton) e, DreamTypes.Lucid.getId()));
        binding.tglDjRecurring.setOnClickListener(e -> toggleSpecialDreamType((ToggleButton) e, DreamTypes.Recurring.getId()));
        binding.tglDjFalseAwakening.setOnClickListener(e -> toggleSpecialDreamType((ToggleButton) e, DreamTypes.FalseAwakening.getId()));

        binding.sldDjDreamMood.setLabelFormatter(value -> dreamMoodLabels[(int) value]);
        binding.sldDjSleepQuality.setLabelFormatter(value -> sleepQualityLabels[(int) value]);
        binding.sldDjDreamClarity.setLabelFormatter(value -> dreamClarityLabels[(int) value]);

        binding.sldDjDreamMood.addOnChangeListener((slider, value, fromUser) -> {
            handleIconSlider((int) value, dreamMoods);
            entry.journalEntry.moodId = DreamMoods.values()[(int) value].getId();
        });
        binding.sldDjSleepQuality.addOnChangeListener((slider, value, fromUser) -> {
            handleIconSlider((int) value, sleepQualities);
            entry.journalEntry.qualityId = SleepQuality.values()[(int) value].getId();
        });
        binding.sldDjDreamClarity.addOnChangeListener((slider, value, fromUser) -> {
            handleIconSlider((int) value, dreamClarities);
            entry.journalEntry.clarityId = DreamClarity.values()[(int) value].getId();
        });

        binding.btnDjCloseEditor.setOnClickListener(e -> {
            if (mBackButtonListener != null) {
                mBackButtonListener.onEvent();
            }
        });

        binding.btnDjBackToText.setOnClickListener(e -> {
            if(mBackButtonListener != null) {
                mBackButtonListener.onEvent();
            }
        });

        binding.btnDjDoneRating.setOnClickListener(e -> {
            if(TextUtils.isEmpty(entry.journalEntry.title) || (TextUtils.isEmpty(entry.journalEntry.description) && entry.audioLocations.isEmpty())) {
                new MaterialAlertDialogBuilder(getContext(), R.style.Theme_LucidSourceKit_ThemedDialog)
                        .setTitle("No content")
                        .setMessage("You have to provide a title and a description or audio recording for your dream journal entry!")
                        .setPositiveButton(getResources().getString(R.string.ok), null)
                        .show();
                return;
            }

            if (mDoneButtonListener != null) {
                mDoneButtonListener.onEvent();
            }
        });
    }

    private void toggleSpecialDreamType(ToggleButton button, String dreamType) {
        if (button.isChecked()) {
            entry.addDreamType(dreamType);
        }
        else {
            entry.removeDreamType(dreamType);
        }
    }

    private void handleIconSlider(int value, ImageView[] icons) {
        if(value < icons.length) {
            for (int i = 0; i < icons.length; i++) {
                if(icons[i] != null) {
                    icons[i].setImageTintList(i == value ? selectedIconColor : unselectedIconColor);
                    LinearLayout.LayoutParams lParams = (LinearLayout.LayoutParams) icons[i].getLayoutParams();
                    lParams.topMargin = i == value ? Tools.dpToPx(getContext(), 4) : 0;
                    lParams.height = i == value ? Tools.dpToPx(getContext(), 20) : Tools.dpToPx(getContext(), 16);
                    icons[i].setLayoutParams(lParams);
                    icons[i].invalidate();
                }
            }
        }
    }

    private void restoreStoredEntry() {
        binding.sldDjDreamMood.setValue(DreamMoods.getEnum(entry.journalEntry.moodId).ordinal());
        handleIconSlider((int) binding.sldDjDreamMood.getValue(), dreamMoods);
        binding.sldDjDreamClarity.setValue(DreamClarity.getEnum(entry.journalEntry.clarityId).ordinal());
        handleIconSlider((int) binding.sldDjDreamClarity.getValue(), dreamClarities);
        binding.sldDjSleepQuality.setValue(SleepQuality.getEnum(entry.journalEntry.qualityId).ordinal());
        handleIconSlider((int) binding.sldDjSleepQuality.getValue(), sleepQualities);

        binding.tglDjNightmare.setChecked(entry.hasSpecialType(DreamTypes.Nightmare.getId()));
        toggleSpecialDreamType(binding.tglDjNightmare, DreamTypes.Nightmare.getId());
        binding.tglDjParalysis.setChecked(entry.hasSpecialType(DreamTypes.SleepParalysis.getId()));
        toggleSpecialDreamType(binding.tglDjParalysis, DreamTypes.SleepParalysis.getId());
        binding.tglDjLucid.setChecked(entry.hasSpecialType(DreamTypes.Lucid.getId()));
        toggleSpecialDreamType(binding.tglDjLucid, DreamTypes.Lucid.getId());
        binding.tglDjRecurring.setChecked(entry.hasSpecialType(DreamTypes.Recurring.getId()));
        toggleSpecialDreamType(binding.tglDjRecurring, DreamTypes.Recurring.getId());
        binding.tglDjFalseAwakening.setChecked(entry.hasSpecialType(DreamTypes.FalseAwakening.getId()));
        toggleSpecialDreamType(binding.tglDjFalseAwakening, DreamTypes.FalseAwakening.getId());
    }

    @Override
    public void onResume() {
        super.onResume();
        getActivity().setRequestedOrientation(ActivityInfo.SCREEN_ORIENTATION_PORTRAIT);
    }

    @Override
    public void onPause() {
        super.onPause();
        getActivity().setRequestedOrientation(ActivityInfo.SCREEN_ORIENTATION_FULL_SENSOR);
    }

    @Override
    public void onDestroyView() {
        compositeDisposable.clear();
        super.onDestroyView();
    }

    public void updatePreview() {
        String title = entry.journalEntry.title;
        String description = entry.journalEntry.description;
        boolean emptyTitle = title == null || title.isEmpty();
        boolean emptyDescription = description == null || description.isEmpty();

        binding.txtJournalTitle.setTextColor(Tools.getAttrColor(emptyTitle ? R.attr.tertiaryTextColor : R.attr.primaryTextColor, getContext().getTheme()));
        binding.txtJournalDescription.setTextColor(Tools.getAttrColor(emptyDescription ? R.attr.tertiaryTextColor : R.attr.secondaryTextColor, getContext().getTheme()));

        binding.txtJournalTitle.setText(emptyTitle ? "No title provided" : title);
        binding.txtJournalDescription.setText(emptyDescription ? "No description provided. Try to write the dream down to better memorize them." : description);

        binding.llTagsHolder.removeAllViews();
        RecyclerViewAdapterDreamJournal.MainViewHolder.setTagList(binding.llTagsHolder, 0, 0, entry.getStringTags(), getActivity());
    }

    public interface OnDreamJournalEntrySaved {
        void onEvent(int entryId);
    }

    public void setJournalEntryId(DreamJournalEntry entry) {
        this.entry = entry;
    }

    public interface OnBackButtonClicked {
        void onEvent();
    }

    public void setOnBackButtonClicked(OnBackButtonClicked eventListener) {
        mBackButtonListener = eventListener;
    }

    public interface OnDoneButtonClicked {
        void onEvent();
    }

    public void setOnDoneButtonClicked(OnDoneButtonClicked eventListener) {
        mDoneButtonListener = eventListener;
    }

    public interface OnCloseButtonClicked {
        void onEvent();
    }

    public void setOnCloseButtonClicked(OnCloseButtonClicked eventListener) {
        mCloseButtonClicked = eventListener;
    }
}

File: app/src/main/java/com/bitflaker/lucidsourcekit/main/dreamjournal/views/DreamJournalEditorRatingView.kt
package com.bitflaker.lucidsourcekit.main.dreamjournal.views

import android.annotation.SuppressLint
import android.content.pm.ActivityInfo
import android.content.res.ColorStateList
import android.os.Bundle
import android.text.TextUtils
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.ImageView
import android.widget.LinearLayout
import android.widget.ToggleButton
import androidx.core.view.updateLayoutParams
import androidx.fragment.app.Fragment
import com.bitflaker.lucidsourcekit.R
import com.bitflaker.lucidsourcekit.main.dreamjournal.rating.DreamClarity
import com.bitflaker.lucidsourcekit.main.dreamjournal.rating.DreamMoods
import com.bitflaker.lucidsourcekit.main.dreamjournal.rating.DreamTypes
import com.bitflaker.lucidsourcekit.main.dreamjournal.rating.SleepQuality
import com.bitflaker.lucidsourcekit.database.dreamjournal.entities.results.DreamJournalEntry
import com.bitflaker.lucidsourcekit.databinding.FragmentJournalEditorRatingBinding
import com.bitflaker.lucidsourcekit.utils.attrColor
import com.bitflaker.lucidsourcekit.utils.attrColorStateList
import com.bitflaker.lucidsourcekit.utils.dpToPx
import com.google.android.material.dialog.MaterialAlertDialogBuilder
import java.security.InvalidParameterException

class DreamJournalEditorRatingView(val entry: DreamJournalEntry) : Fragment() {
    private val dreamMoodLabels = arrayOf("Terrible", "Poor", "Okay", "Great", "Outstanding")
    private val sleepQualityLabels = arrayOf("Terrible", "Poor", "Great", "Outstanding")
    private val dreamClarityLabels = arrayOf("Very Cloudy", "Cloudy", "Clear", "Crystal Clear")
    private lateinit var binding: FragmentJournalEditorRatingBinding
    private lateinit var dreamMoods: Array<ImageView>
    private lateinit var sleepQualities: Array<ImageView>
    private lateinit var dreamClarities: Array<ImageView>
    private lateinit var unselectedIconColor: ColorStateList
    private lateinit var selectedIconColor: ColorStateList
    var onBackButtonListener: (() -> Unit)? = null
    var onDoneButtonListener: (() -> Unit)? = null

    override fun onCreateView(inflater: LayoutInflater, container: ViewGroup?, savedInstanceState: Bundle?): View? {
        binding = FragmentJournalEditorRatingBinding.inflate(inflater, container, false)
        return binding.root
    }

    override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
        super.onViewCreated(view, savedInstanceState)

        val context = requireContext()
        unselectedIconColor = context.attrColorStateList(R.attr.colorSurfaceContainerHigh)
        selectedIconColor = context.attrColorStateList(R.attr.colorOnSurface)

        // Initialize indicator image arrays
        dreamMoods = arrayOf(
            binding.imgVeryDissatisfied,
            binding.imgDissatisfied,
            binding.imgNeutralSatisfied,
            binding.imgSatisfied,
            binding.imgVerySatisfied
        )
        sleepQualities = arrayOf(
            binding.imgVeryBadQuality,
            binding.imgBadQuality,
            binding.imgGoodQuality,
            binding.imgVeryGoodQuality
        )
        dreamClarities = arrayOf(
            binding.imgVeryUnclear,
            binding.imgUnclear,
            binding.imgClear,
            binding.imgVeryClear
        )

        restoreStoredEntry()

        // Set handlers for toggling special dream types
        binding.tglDjNightmare.setOnClickListener { e ->
            toggleSpecialDreamType(e as ToggleButton, DreamTypes.Nightmare.id)
        }
        binding.tglDjParalysis.setOnClickListener { e ->
            toggleSpecialDreamType(e as ToggleButton, DreamTypes.SleepParalysis.id)
        }
        binding.tglDjLucid.setOnClickListener { e ->
            toggleSpecialDreamType(e as ToggleButton, DreamTypes.Lucid.id)
        }
        binding.tglDjRecurring.setOnClickListener { e ->
            toggleSpecialDreamType(e as ToggleButton, DreamTypes.Recurring.id)
        }
        binding.tglDjFalseAwakening.setOnClickListener { e ->
            toggleSpecialDreamType(e as ToggleButton, DreamTypes.FalseAwakening.id)
        }

        // Set label formatter values
        binding.sldDjDreamMood.setLabelFormatter { value -> dreamMoodLabels[value.toInt()] }
        binding.sldDjSleepQuality.setLabelFormatter { value -> sleepQualityLabels[value.toInt()] }
        binding.sldDjDreamClarity.setLabelFormatter { value -> dreamClarityLabels[value.toInt()] }

        // Set handlers for slider value changes
        binding.sldDjDreamMood.addOnChangeListener { _, value, fromUser ->
            handleIconSlider(value.toInt(), dreamMoods)
            entry.journalEntry.moodId = DreamMoods.entries[value.toInt()].id
        }
        binding.sldDjSleepQuality.addOnChangeListener { _, value, fromUser ->
            handleIconSlider(value.toInt(), sleepQualities)
            entry.journalEntry.qualityId = SleepQuality.entries[value.toInt()].id
        }
        binding.sldDjDreamClarity.addOnChangeListener { _, value, fromUser ->
            handleIconSlider(value.toInt(), dreamClarities)
            entry.journalEntry.clarityId = DreamClarity.entries[value.toInt()].id
        }

        // Set handler for closing editor
        binding.btnDjCloseEditor.setOnClickListener {
            onBackButtonListener?.invoke()
        }

        // Set handler for going back to the text editor
        binding.btnDjBackToText.setOnClickListener {
            onBackButtonListener?.invoke()
        }

        // Set handler for saving entry
        binding.btnDjDoneRating.setOnClickListener {
            val hasMissingTitle = TextUtils.isEmpty(entry.journalEntry.title)
            val hasMissingContent = (TextUtils.isEmpty(entry.journalEntry.description) && entry.audioLocations.isEmpty())
            if (hasMissingTitle || hasMissingContent) {
                MaterialAlertDialogBuilder(context, R.style.Theme_LucidSourceKit_ThemedDialog)
                    .setTitle("Missing fields")
                    .setMessage("You have to provide a title and a description or audio recording for your dream journal entry!")
                    .setPositiveButton(resources.getString(R.string.ok), null)
                    .show()
                return@setOnClickListener
            }
            onDoneButtonListener?.invoke()
        }
    }

    private fun toggleSpecialDreamType(button: ToggleButton, dreamType: String) {
        if (button.isChecked) {
            entry.addDreamType(dreamType)
        } else {
            entry.removeDreamType(dreamType)
        }
    }

    private fun handleIconSlider(value: Int, icons: Array<ImageView>) {
        if (value >= icons.size) {
            throw InvalidParameterException("Value exceeded icon count")
        }

        // Update the selected state of all icons
        for (i in icons.indices) {
            val isSelected = i == value
            icons[i].setImageTintList(if (isSelected) selectedIconColor else unselectedIconColor)
            icons[i].updateLayoutParams<LinearLayout.LayoutParams> {
                topMargin = if (isSelected) 4.dpToPx else 0
                height = if (isSelected) 20.dpToPx else 16.dpToPx
            }
            icons[i].invalidate()
        }
    }

    private fun restoreStoredEntry() {
        // Restore slider states and values
        binding.sldDjDreamMood.value = DreamMoods.getEnum(entry.journalEntry.moodId).ordinal.toFloat()
        handleIconSlider(binding.sldDjDreamMood.value.toInt(), dreamMoods)
        binding.sldDjDreamClarity.value = DreamClarity.getEnum(entry.journalEntry.clarityId).ordinal.toFloat()
        handleIconSlider(binding.sldDjDreamClarity.value.toInt(), dreamClarities)
        binding.sldDjSleepQuality.value = SleepQuality.getEnum(entry.journalEntry.qualityId).ordinal.toFloat()
        handleIconSlider(binding.sldDjSleepQuality.value.toInt(), sleepQualities)

        // Restore toggle button states
        binding.tglDjNightmare.setChecked(entry.hasSpecialType(DreamTypes.Nightmare.id))
        toggleSpecialDreamType(binding.tglDjNightmare, DreamTypes.Nightmare.id)
        binding.tglDjParalysis.setChecked(entry.hasSpecialType(DreamTypes.SleepParalysis.id))
        toggleSpecialDreamType(binding.tglDjParalysis, DreamTypes.SleepParalysis.id)
        binding.tglDjLucid.setChecked(entry.hasSpecialType(DreamTypes.Lucid.id))
        toggleSpecialDreamType(binding.tglDjLucid, DreamTypes.Lucid.id)
        binding.tglDjRecurring.setChecked(entry.hasSpecialType(DreamTypes.Recurring.id))
        toggleSpecialDreamType(binding.tglDjRecurring, DreamTypes.Recurring.id)
        binding.tglDjFalseAwakening.setChecked(entry.hasSpecialType(DreamTypes.FalseAwakening.id))
        toggleSpecialDreamType(binding.tglDjFalseAwakening, DreamTypes.FalseAwakening.id)
    }

    fun updatePreview() {
        val context = requireContext()
        val title = entry.journalEntry.title
        val description = entry.journalEntry.description
        val emptyTitle = title.isNullOrEmpty()
        val emptyDescription = description.isNullOrEmpty()

        // Set text colors for title and description
        binding.txtJournalTitle.setTextColor(context.attrColor(if (emptyTitle) R.attr.tertiaryTextColor else R.attr.primaryTextColor))
        binding.txtJournalDescription.setTextColor(context.attrColor(if (emptyDescription) R.attr.tertiaryTextColor else R.attr.secondaryTextColor))

        // Set text values for title and description
        binding.txtJournalTitle.text = if (emptyTitle) "No title provided" else title
        binding.txtJournalDescription.text = if (emptyDescription) "No description provided. Try to write the dream down to better memorize them." else description

        // Load tags container into preview
        binding.llTagsHolder.removeAllViews()

        // TODO: Rework the way this preview section works / put something different up there
//        RecyclerViewAdapterDreamJournal.MainViewHolder.setTagList(
//            binding.llTagsHolder,
//            0,
//            0,
//            entry.stringTags,
//            activity
//        )
    }

    @SuppressLint("SourceLockedOrientationActivity")
    override fun onResume() {
        super.onResume()
        requireActivity().setRequestedOrientation(ActivityInfo.SCREEN_ORIENTATION_PORTRAIT)
    }

    override fun onPause() {
        super.onPause()
        requireActivity().setRequestedOrientation(ActivityInfo.SCREEN_ORIENTATION_FULL_SENSOR)
    }
}